Topic content filtered topic

iuhilnehc-ynos opened the pull request:
Just for internal review

Commit:
to support a feature of content filtered topic

Signed-off-by: Chen Lihui <lihui.chen@sony.com>

Commit:
Add interface without implementation. TODO

Signed-off-by: Chen Lihui <lihui.chen@sony.com>

Commit:
update for set_cft_expression_parameters and get_cft_expression_parameters

Signed-off-by: Chen Lihui <lihui.chen@sony.com>

Commit:
fix memory leak

Signed-off-by: Chen Lihui <lihui.chen@sony.com>

Commit:
cft topic name use fqn+"_ContentFilterTopic"

Signed-off-by: Chen Lihui <lihui.chen@sony.com>

Commit:
to support some complicated cases

Signed-off-by: Chen Lihui <lihui.chen@sony.com>

Commit:
add a check statement if current subscriber not support cft

Signed-off-by: Chen Lihui <lihui.chen@sony.com>

Commit:
update comments

Signed-off-by: Chen Lihui <lihui.chen@sony.com>

Commit:
fix mem leak

Signed-off-by: Chen Lihui <lihui.chen@sony.com>

Commit:
remove unnecessary comments

Signed-off-by: Chen Lihui <lihui.chen@sony.com>

fujitatomoya commented:
CC: @Barry-Xu-2018

fujitatomoya commented on rmw_connextdds/src/rmw_api_impl_ndds.cpp:
```diff
@@ -742,6 +742,28 @@ rmw_subscription_get_actual_qos(
 
 
+rmw_ret_t
+rmw_subscription_set_cft_expression_parameters(
```
just a comment, as far as i see this is the right place to implement (non-dds) rmw interfaces.

fujitatomoya commented on rmw_connextddsmicro/src/rmw_api_impl_rtime.cpp:
```diff
@@ -742,6 +742,28 @@ rmw_subscription_get_actual_qos(
 }
 
 
+rmw_ret_t
```
I do not really think connext micro supports the contentfilteredtopic, probably we can delete these or return NOT_SUOPPORT.

iuhilnehc-ynos replied:
Actually, I thought micro does need this feature. OK, I'll update it.

fujitatomoya commented on rmw_connextdds_common/include/rmw_connextdds/rmw_impl.hpp:
```diff
@@ -1059,22 +1071,35 @@ class RMW_Connext_Subscriber
   rmw_gid_t ros_gid;
   const bool created_topic;
-  RMW_Connext_SubscriberStatusCondition status_condition;
+  std::shared_ptr<RMW_Connext_SubscriberStatusCondition> status_condition;
```
just a comment. this needs to be shared_ptr so that after reset (reset filter expression and expression parameter for new data reader) deallocation happens.

fujitatomoya commented on rmw_connextdds_common/include/rmw_connextdds/rmw_impl.hpp:
```diff
@@ -1059,22 +1071,35 @@ class RMW_Connext_Subscriber
 
+  void set_node(const rmw_node_t * node)
+  {
+    this->node = node;
```
just a comment, this is required to update the node graph when resetting contentfilteredtopic.

fujitatomoya commented on rmw_connextdds_common/include/rmw_connextdds/rmw_impl.hpp:
```diff
@@ -1059,22 +1071,35 @@ class RMW_Connext_Subscriber
   const bool internal;
 
 private:
+  enum class TOPIC_TYPE {NORMAL, CFT};
+  rmw_ret_t reset_topic(TOPIC_TYPE type, const char * filter_expression, const rcutils_string_array_t * expression_parameters);
```
what are those? never used?

iuhilnehc-ynos replied:
something I'll use while refactoring code. Yes, it's not used, I will remove it

fujitatomoya commented on rmw_connextdds_common/include/rmw_connextdds/rmw_impl.hpp:
```diff
@@ -1059,22 +1071,35 @@ class RMW_Connext_Subscriber
   size_t loan_len;
   size_t loan_next;
   std::mutex loan_mutex;
+  std::mutex cft_mutex;
```
not used?

iuhilnehc-ynos replied:
remove it

iuhilnehc-ynos replied:
Oh, It will be used in the set/get cft operations.

fujitatomoya commented on rmw_connextdds_common/include/rmw_connextdds/rmw_impl.hpp:
```diff
@@ -1059,22 +1071,35 @@ class RMW_Connext_Subscriber
+  std::mutex cft_mutex;
+  const rmw_node_t * node;
+  std::string fqtopic_name;
+  rmw_qos_profile_t qos_policies;
```
just a comment. this needs to be save in this class, since resetting.

fujitatomoya commented on rmw_connextdds_common/include/rmw_connextdds/rmw_impl.hpp:
```diff
@@ -1059,22 +1071,35 @@ class RMW_Connext_Subscriber
+  const rmw_node_t * node;
+  std::string fqtopic_name;
+  rmw_qos_profile_t qos_policies;
+  bool ignore_local;
```
just a comment. also for resetting, needs to be saved.

fujitatomoya commented on rmw_connextdds_common/src/common/rmw_impl.cpp:
```diff
@@ -1748,11 +1749,13 @@ RMW_Connext_Subscriber::create(
 
   DDS_TopicDescription * sub_topic = DDS_Topic_as_topicdescription(topic);
-
-  if (nullptr != cft_name) {
```
especially here, we need to consult with RTI, to discuss on the CFT interfaces.

fujitatomoya commented on rmw_connextdds_common/src/common/rmw_impl.cpp:
```diff
@@ -1748,11 +1749,13 @@ RMW_Connext_Subscriber::create(
   DDS_TopicDescription * sub_topic = DDS_Topic_as_topicdescription(topic);
-
-  if (nullptr != cft_name) {
+  std::string cft_topic_name = fqtopic_name+"_ContentFilterTopic";
```
this `cft_topic_name` will not be printed via `ros2cli`, right? as long as i can see, parent topic is constructed with `topic` so that parent topic name will be printed.

iuhilnehc-ynos replied:
Yes

fujitatomoya commented on rmw_connextdds_common/src/common/rmw_impl.cpp:
```diff
@@ -1841,6 +1844,9 @@ RMW_Connext_Subscriber::create(
     RMW_CONNEXT_LOG_ERROR_SET("failed to allocate RMW subscriber")
     return nullptr;
   }
+  rmw_sub_impl->fqtopic_name = fqtopic_name;
+  rmw_sub_impl->qos_policies = *qos_policies;
```
just a comment, save them for resetting.

fujitatomoya commented on rmw_connextdds_common/src/common/rmw_impl.cpp:
```diff
@@ -1748,11 +1749,13 @@ RMW_Connext_Subscriber::create(
-
-  if (nullptr != cft_name) {
+  std::string cft_topic_name = fqtopic_name+"_ContentFilterTopic";
+  if (nullptr != subscriber_options->filter_expression) {
```
how about subscriber_options->filter_expression is not null but `""`? in that case, i think we do not need to issue `rmw_connextdds_create_contentfilteredtopic` if i am not mistaken.

iuhilnehc-ynos replied:
Thank you. :+1: 
I'll add this check here.

fujitatomoya commented on rmw_connextdds_common/src/common/rmw_impl.cpp:
```diff
@@ -2031,6 +2037,330 @@ RMW_Connext_Subscriber::take_serialized(
+    // case 2: create a new cft if there is no cft topic and filter_expression is not empty("")
+    // remove old reader
+    if (RMW_RET_OK !=
+      rmw_connextdds_graph_on_subscriber_deleted(
+        ctx, node, this))
```
if i am not mistaken, this is required since we will re-create dds reader, right?

iuhilnehc-ynos replied:
Yes

fujitatomoya commented on rmw_connextdds_common/src/common/rmw_impl.cpp:
```diff
@@ -2031,6 +2037,330 @@ RMW_Connext_Subscriber::take_serialized(
+    this->status_condition->invalidate();
+
+    if (DDS_RETCODE_OK !=
+    DDS_Subscriber_delete_datareader(this->dds_subscriber(), this->dds_reader))
```
```suggestion
      DDS_Subscriber_delete_datareader(this->dds_subscriber(), this->dds_reader))
```

fujitatomoya commented on rmw_connextdds_common/src/common/rmw_impl.cpp:
```diff
@@ -2031,6 +2037,330 @@ RMW_Connext_Subscriber::take_serialized(
 }
 
+rmw_ret_t
+RMW_Connext_Subscriber::set_cft_expression_parameters(
```
I think that it would be probably required mutex lock for set/get operation?

fujitatomoya reviewed (commented):
overall looks good, i left mostly as `just a comment` you can just resolve it. but also some questions, please have them checked.

Commit:
to support reset content filtered topic with empty string.

remove constness for rmw_subscription because is_cft_supported might be updated

Signed-off-by: Chen Lihui <lihui.chen@sony.com>

iuhilnehc-ynos reviewed (commented):
Thanks for your comment, I'll remove some unuseful source code and update `micro` part.

Barry-Xu-2018 commented on rmw_connextdds_common/src/common/rmw_event.cpp:
```diff
@@ -124,12 +124,14 @@ rmw_api_connextdds_take_event(
+    }
+
+    *taken = true;
   }
```
Whether to show error while condition is nullptr?
This is an abnormal case. May help to address the issue via error log.

iuhilnehc-ynos replied:
I'll add a warning to show the condition was reset

Barry-Xu-2018 commented on rmw_connextdds_common/src/common/rmw_impl.cpp:
```diff
@@ -2031,6 +2037,330 @@ RMW_Connext_Subscriber::take_serialized(
+    // that means reset
+    if (filter_expression && *filter_expression == '\0') {
+      // remove old reader
+      if (RMW_RET_OK !=
+        rmw_connextdds_graph_on_subscriber_deleted(
+          ctx, node, this))
+      {
+        RMW_CONNEXT_LOG_ERROR("failed to update graph for subscriber")
+        return RMW_RET_ERROR;
+      }
+
+      this->status_condition->invalidate();
+
+      if (DDS_RETCODE_OK !=
+      DDS_Subscriber_delete_datareader(this->dds_subscriber(), this->dds_reader))
+      {
+        RMW_CONNEXT_LOG_ERROR_SET(
+          "failed to delete DDS DataReader")
+        return RMW_RET_ERROR;
+      }
```
These codes are the same as line 2054~2070.   
Is it possible to avoid it ?

iuhilnehc-ynos replied:
They're different cases for the non-cft topic reader(I called it normal non-cft) and the cft topic reader.
We can combine them with one if statement, but it seems not clear. I'll try to update them.

Barry-Xu-2018 commented on rmw_connextdds_common/src/common/rmw_impl.cpp:
```diff
@@ -2031,6 +2037,330 @@ RMW_Connext_Subscriber::take_serialized(
+            ctx, node, this))
+        {
+          RMW_CONNEXT_LOG_ERROR("failed to update graph for subscriber")
+          return RMW_RET_ERROR;
```
In this function, most of 'RMW_RET_ERROR' means the critical error.     
Maybe it better to add some descriptions on that user have better to destroy and re-create subscriber while getting return value 'RMW_RET_ERROR'.

iuhilnehc-ynos replied:
I used the same destroy source code from some destroy function to delete these resources. If it failed any place on this function, I don't think the user can use some correct function to destroy this subscriber.

Barry-Xu-2018 replied:
Get it.

Commit:
address review

Co-authored-by: Barry Xu <barry.xu@sony.com>
Co-authored-by: Tomoya.Fujita <Tomoya.Fujita@sony.com>
Signed-off-by: Chen Lihui <lihui.chen@sony.com>

Commit:
not to reset subscriber if message loaned

Signed-off-by: Chen Lihui <lihui.chen@sony.com>

Commit:
use atomic for loan_len

Signed-off-by: Chen Lihui <lihui.chen@sony.com>

Commit:
nit

Signed-off-by: Chen Lihui <lihui.chen@sony.com>

Commit:
use make_scope_exit

Signed-off-by: Chen Lihui <lihui.chen@sony.com>
